url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/Examples/Dvfs_example_more_complex.java

```diff
@@ -0,0 +1,609 @@
+	}
+
+
+	/**
+	 * Prints the Cloudlet objects.
+	 *
+	 * @param list list of Cloudlets
+	 */
+	private static void printCloudletList(List<Cloudlet> list) {
+		int size = list.size();
+		Cloudlet cloudlet;
+
+		String indent = "\t";
+		Log.printLine();
+		Log.printLine("========== OUTPUT ==========");
+		Log.printLine("Cloudlet ID" + indent + "STATUS" + indent
+				+ "Resource ID" + indent + "VM ID" + indent + "Time" + indent
+				+ "Start Time" + indent + "Finish Time");
+
+		DecimalFormat dft = new DecimalFormat("###.##");
+		for (int i = 0; i < size; i++) {
+			cloudlet = list.get(i);
+			Log.print(indent + cloudlet.getCloudletId());
+
+			if (cloudlet.getCloudletStatus() == Cloudlet.SUCCESS) {
+				Log.printLine(indent + "SUCCESS"
+					+ indent + indent + cloudlet.getResourceId()
+					+ indent + cloudlet.getVmId()
+					+ indent + dft.format(cloudlet.getActualCPUTime())
+					+ indent + dft.format(cloudlet.getExecStartTime())
+					+ indent + indent + dft.format(cloudlet.getFinishTime())
+				);
+			}
+		}
+	}
+
+
+
+public static class GlobalBroker extends SimEntity {
+
+/*
+
+This GlobalBroker class is used to create successively broker that launch one or multiple VM. Indeed, a broker can "contain" one or more VM.
+One (and only one) cloudlet is attached to one VM. The reasone is if we use only one VM, this has to have to the same size as the Host
+to be able to generate 100% cpu load.
+
+In our simulation we want to generate a pre-defined scenario, with multiple predefine phase.
+It's more convinient to create different type of VM and different type of Cloudlet.
+By mixing these different type of VM/Cloudlet we can easaly manage the CPU load of the simulation.
+The size of the VM define the grain with we are able to increase/decrease the CPU load, and the Cloudlet length determine the duration of its execution,
+depend on which type of VM it is launched.
+
+
+DVFS
+----
+
+This scenario is defined at Maximum frequency. So all broker is set to have the good CPU load curve at Maximum frequency.
+When the CPU is decrease, the execution of VM/CloudLet could be slowed and the sequence of brokers must take into account this slowdown.
+If we create broker with STATIC delay start, it could be works at Maximum frequency, but not at all for another (lower) frequency.
+
+So, the launching of a broker could be create dynamically at the end of another broker. (Event Post Broker)
+
+*/
+
+
+
+		private static final int STEP_1 = 0;
+		private static final int STEP_2 = 1;
+		private static final int STEP_3 = 2;
+		private static final int STEP_4 = 3;
+		
+		private List<Vm> vmList_=null;
+		private List<Cloudlet> cloudletList_=null;
+		private List<PowerDatacenterBroker> created_brokers=Collections.synchronizedList(new ArrayList<PowerDatacenterBroker>());
+		PowerDatacenterBroker broker;
+		
+		int no_broker = 1;
+
+		
+		
+		public GlobalBroker(String name) {
+			super(name);
+		}
+
+		@Override
+		public void processEvent(SimEvent ev) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 64 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.